lib/puppet/resource_api/property.rb

```diff
@@ -69,10 +86,40 @@ def rs_value
   # method overloaded only for the :ensure property, add option to check if the
   # rs_value matches is. Only if the class is child of
   # Puppet::ResourceApi::Property.
-  def def_insync?
+  def def_ensure_insync?
     define_singleton_method(:insync?) { |is| rs_value.to_s == is.to_s }
   end
 
+  def def_custom_insync?
+    define_singleton_method(:insync?) do |is|
+      provider    = @referrable_type.my_provider
+      context     = @referrable_type.context
+      should_hash = @resource.rsapi_canonicalized_target_state
+      is_hash     = @resource.rsapi_current_state
+      title       = @resource.rsapi_title
+
+      raise(Puppet::DevError, 'No insync? method defined in the provider; an insync? method must be defined if the custom_insync feature is defined for the type') unless provider.respond_to?(:insync?)
+      provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash)
```

Comment:
Since `insync?` is an extension point, it seems likely that it could raise. Is any additional exception handling needed here to ensure the resource is marked as failed, a `Puppet::Resource::Status` created for the resource in the report, the error is captured in the log, including backtrace info (when running with `--trace`)?

Reply:
I can add some testing to validate behavior for when `insync?` raises and if it doesn't work as expected, we can refactor to be sure that it does.

Reply:
Testing added and behavior validated.

Reply:
tests are in https://github.com/puppetlabs/puppet-resource_api/pull/285/files#diff-08381f7c9c76cb719d8c29f2553a09ea03a0505c24fac62b51d03b6af50c6b5fR13

Reply:
:rofl: 5 seconds apart :+1: